[1단계 - 자동차 경주 구현] 파랑(이하은) 미션 제출합니다.

summerlunaa opened the pull request:
안녕하세요 파랑입니다!

코딩하면서 궁금했던 점이나 해결못한 부분들이 있어 질문남깁니다.

1. 랜덤값에 따라 자동차 경주의 결과가 다르게 나와서 테스트 코드를 짜는 데 어려움을 겪었습니다. 시간이 촉박해서 이 부분은 우선 구현하지 않았는데 어떻게 구현할 수 있을까요? 궁금합니다!

2. 저는 이번에 테스트 코드를 처음 작성해보았는데 테스트를 통과하려다보니 테스트만을 위한 메서드를 만들어야 하는 경우도 생기더라구요. 원래 테스트만을 위한 메서드를 구현하는 방식으로 작업하는 건가요?

3. 패키지 안에 들어갈 클래스가 하나이더라도 패키지를 만드는 것이 좋을까요?

4. 아래의 경우 이름을 `splitAndTrim`으로 짓는 것이 좋을까요? 하나의 메서드가 하나의 일만 하도록 구현하라고 하셨는데 이런 경우엔 메서드를 어떻게 나눠야 할지, 이름을 어떻게 지어야 할지 궁금합니다!
```java
public static String[] splitAndTrim(String names) {
		String[] splitNames = names.split(STRING_REGEX);
		Arrays.setAll(splitNames, i -> splitNames[i].trim());
		return splitNames;
	}
```

특히나 테스트 코드가 처음이라 제 코드에 확신이 서질 않습니다ㅠㅠ 조언 많이 해주시면 감사하겠습니다!!!
부족한 코드인 것 같아서 부끄럽지만 잘 부탁드립니다!

summerlunaa committed:
test : splitTest 구현

summerlunaa committed:
test : removeBracketTest 구현

summerlunaa committed:
test : getCharTest 구현

summerlunaa committed:
test : 불필요한 import 삭제

summerlunaa committed:
test : sizeTest 구현

summerlunaa committed:
test : containsTest 구현

summerlunaa committed:
test : containsFalseTest 구현

summerlunaa committed:
feat : 빈 문자열 또는 null 값을 입력할 경우 0을 반환하는기능 구현

summerlunaa committed:
feat : 숫자 하나를 문자열로 입력할 경우 해당 숫자를 반환하는 기능 구현

summerlunaa committed:
feat : 숫자 두 개를 컴마(,) 구분자로 입력할 경우 두 숫자의 합을 반환하는 기능 구현

summerlunaa committed:
feat : 구분자를 컴마(,) 이외에 콜론(:)을 사용할 수 있도록 기능 구현

summerlunaa committed:
feat : 커스텀 구분자를 지정하여 사용할 수 있도록 기능 구현

summerlunaa committed:
feat : 음수인 경우 RuntimeException 예외 발생시키는 기능 구현

summerlunaa committed:
feat : 숫자 외의 입력이 들어온 경우 예외를 발생시키도록 기능 구현

summerlunaa committed:
docs : 구현할 기능 목록 작성

summerlunaa committed:
feat : 자동차 이름 입력에 대한 검증 구현

summerlunaa committed:
test : 자동차 이름 입력 검증에 대한 test 코드 작성

summerlunaa committed:
test : 이동횟수 입력 검증에 대한 test 코드 작성

summerlunaa committed:
feat : 이동횟수 입력에 대한 검증 구현

summerlunaa committed:
test : Car 객체에 대한 테스트 작성

summerlunaa committed:
feat : Car 객체 및 메서드 구현

summerlunaa committed:
test : Cars 객체에 대한 테스트 작성

summerlunaa committed:
feat : Cars 객체 및 메서드 구현

summerlunaa committed:
feat : 자동차 이름을 입력받는 getCarNames 구현

summerlunaa committed:
feat : 이동횟수를 입력받는 getMoveCount 구현

summerlunaa committed:
feat : 각 레이스의 결과를 출력하는 printCarPosition 구현, GameController 구현

summerlunaa committed:
feat : car 클래스의 toStirng 구현

summerlunaa committed:
feat : startRace에 출력하는 부분 추가 구현

summerlunaa committed:
chore : 패키지 구조 변경

summerlunaa committed:
chore : 패키지 구조 변경

summerlunaa committed:
feat : splitAndTrim 메서드 구현

summerlunaa committed:
feat : 출력을 위한 OutputView 구현

summerlunaa committed:
feat : Cars 클래스 및 메서드 구현

summerlunaa committed:
feat : main 메서드 구현

summerlunaa committed:
refactor : 하드코딩을 상수로 변경

summerlunaa committed:
refactor : 하드코딩을 상수로 변경

summerlunaa committed:
refactor : StringAddCalculator 메소드 수정 및 분리

summerlunaa committed:
refactor : StringAddCalculatorTest 메소드 이름 수정, empty와 null 테스트 분리

summerlunaa committed:
fix : Utils 상수 오타로인한 오류 수정

summerlunaa committed:
refactor : 하드코딩 상수로 변경

summerlunaa committed:
refactor : 메서드 순서 변경 및 스트림 사용

summerlunaa committed:
refactor : 변수 이름 변경 및 스트림 사용

summerlunaa committed:
test : 메서드 이름 변경

summerlunaa committed:
refactor : 메서드, 변수 이름 변경

summerlunaa committed:
refactor : startRace 메서드 분리

summerlunaa committed:
refactor : 스트림 적용

summerlunaa committed:
style : 자바 코딩 컨벤션 적용

jjj0611 commented on src/test/java/junitstudy/StringTest.java:
```diff
@@ -0,0 +1,40 @@
+		String testString = "abc";
+		char zero = testString.charAt(0);
+		assertThat(zero).isEqualTo('a');
+		assertThatThrownBy(() -> {
```
{}이 불필요하게 들어간 것 같아요! 제거 되어도 되지 않을까요?

jjj0611 commented on src/test/java/junitstudy/StringTest.java:
```diff
@@ -0,0 +1,40 @@
+		String[] result2 = testString2.split(",");
+
+		assertThat(result1).containsExactly("1", "2");
+		assertThat(result2).containsExactly("1");
```
학습테스트이긴 하지만 한 테스트에서 2가지를 동시에 테스트하고 있네요.
하나로 두신 이유가 있을까요?
혹시 두 가지 케이스를 두 개의 테스트로 쪼개면 어떨까요?

jjj0611 commented on src/test/java/racingcar/NameValidatorTest.java:
```diff
@@ -0,0 +1,46 @@
+
+import static racingcar.validator.NameValidator.*;
+
+public class NameValidatorTest {
```
전체적으로 사용하지 않는 bracket은 제거하면 어떨까요? intellij에 회색으로 가이드가 될텐데 intellij의 가이드를 잘 활용한다면 도움이 많이 되실 거에요 :)

summerlunaa replied:
감사합니다! 예시 코드를 참고하면서 쓰다보니 bracket을 제거하지 않고 그대로 쓴 것 같네요! intellij의 가이드 참고하도록 하겠습니다ㅎㅎ

jjj0611 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,35 @@
+	public static final String REQUEST_CAR_NAMES = "경주할 자동차 이름을 입력하세요.(이름은 쉼표(,) 기준으로 구분)";
+	private static final String REQUEST_MOVE_COUNT = "시도할 회수는 몇회인가요?";
+
+	public static String getCarNames(Scanner scanner) {
```
validator를 통해서 입력받은 문자열을 모두 잘라서 각각 확인한 후에 다시 입력 받은 문자열을 그대로 넘겨주고 있습니다.
그러면 뒤쪽에서 다시 나눈 작업을 해야하는데, 너무 중복으로 하고 있는 것이 많지 않을까요?

역할을 분리하기 위해 중복적으로 로직이 들어가는 경우도 있긴하지만, 
최대한 역할을 잘 분리하여 중복 로직을 제거하는 것이 좋을 것 같아요 :)

summerlunaa replied:
InputView에서 문자열을 나누고 반환까지 해주면 너무 하나의 메서드가 일을 많이 하지 않나?생각했는데, 검증 로직을 다른 클래스로 옮기니 이것에 대한 고민이 어느정도 해결된 것 같습니다! 그래서 중복되는 로직을 제거하고 InputView의 getCarNames 메서드에서 문자열 나누는 작업을 한 번만 하도록 수정했습니다!

jjj0611 commented on src/main/java/racingcar/GameController.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Scanner;
+
+public class GameController {
+	private final Scanner scanner;
```
scanner를 application에서 생ㅅ어한 후에 controller에 주입하여 inputview에 넣어주고 있네요.
Application에서 생성하여 주입한 이유가 있을까요?

summerlunaa replied:
계속해서 입력을 받아야하는데 받을 때마다 scanner를 생성하는 것이 비효율적인 것 같아서 GameController에서 한 번만 생성하여 계속 재사용할 수 있도록 했습니다!

jjj0611 commented on src/main/java/racingcar/GameController.java:
```diff
@@ -0,0 +1,31 @@
+
+	public void run() {
+		String carNames = InputView.getCarNames(scanner);
+		Cars cars = new Cars(Utils.splitAndTrim(carNames));
```
NameValidator에서도 이야기 했지만, 이 두 라인에서 이름 검증을 어떻게 거치고 있는지 다시 한 번 고민해보면 좋을 것 같아요.
자동차의 이름이 5글자가 되면 안된다는 것은 자동차의 로직일까요? 입력의 로직일까요?
자동차의 이름이 중복될 수 없다는 것은 경주의 로직일까요? 입력의 로직일까요?

또한 입력 값을 각각 나눠서 각 이름에 대한 검증을 한 뒤에 다시 그 문자열을 받아서 split하고 있는데, 여기서도 중복 로직이 많이 발생하는 것 같아요. 로직의 중복을 최소화해볼까요? :)

summerlunaa replied:
음ㅠㅠ 제가 정확히 이해한 건지 모르겠지만 이름 검증을 InputView에서 할지, Car에서 할지 고민해보라는 말씀이신 거 맞을까요? 프리코스때에도 가장 고민했던 부분이 입력에 대한 검증을 어디서 처리할 것인가였는데 지금까지는 "입력"에 대한 검증이니까 InputView에서 처리해서 넘겨주는 것이 맞지 않을까? 생각해서 계속 InputView에서 처리했던 것 같습니다. 그런데 제이님의 코멘트를 보고 고민해보니 "자동차"의 로직에 검증이 포함된다는 생각이 듭니다!! 그래서 검증하는 부분을 Car 클래스로 이동했는데 괜.. 찮나요!?

jjj0611 commented on src/main/java/racingcar/GameController.java:
```diff
@@ -0,0 +1,31 @@
+	public void run() {
+		String carNames = InputView.getCarNames(scanner);
+		Cars cars = new Cars(Utils.splitAndTrim(carNames));
+		int moveCount = Integer.parseInt(InputView.getMoveCount(scanner));
```
Integer.parseInt가 발생할 수 있는 예외는 어떤게 있을까요?
위와 같이 구현된다면 InputView에서의 검증로직과 매우 강한 결합을 갖게 됩니다.
각 객체가 담당해야하는 역할을 나눠보면 어떨까요?
필요하다면 새로운 객체가 생성되어도 좋을 것 같아요 :)

summerlunaa replied:
parseInt 메서드가 너무 반복적으로 쓰이는 것 같아서 Utils에 strToInt라는 메서드를 새로 만들고 한 번만 변경하도록 바꿨습니다. 그렇게 해서 run 메서드에서 정수로 변경하는 부분을 제거했습니다! 제가 이해한 게 맞을까요?

jjj0611 commented on src/test/java/racingcar/MoveCountValidatorTest.java:
```diff
@@ -0,0 +1,33 @@
+
+import static racingcar.validator.MoveCountValidator.*;
+
+class MoveCountValidatorTest {
```
이 테스트에서도 전체적으로 불필요한 bracket이 있습니다.
제거하는 것이 가독성에 더 좋지 않을까요?

jjj0611 commented on src/test/java/racingcar/MoveCountValidatorTest.java:
```diff
@@ -0,0 +1,33 @@
+			MoveCountValidator.validateMoveCount(moveCount);
+		}).isInstanceOf(IllegalArgumentException.class)
+			.hasMessageContaining(ERROR_NOT_POSITIVE);
+	}
```
엣지 케이스에 대해서 잘 테스트 해주셨네요 👍

jjj0611 commented on src/main/java/racingcar/validator/NameValidator.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public class NameValidator {
```
NameValidator라는 이름을 보면 1개의 Name에 대한 validate를 하는 것으로 생각이 됩니다.
그런데 실제로 validate를 하는 parameter의 이름이 names이고, 내부 로직으로 해당 역할을 이해해야겠네요. 이름에서 좀 더 명확하게 의도가 드러나면 어떨까요?

summerlunaa replied:
CarNameValidator와 RacingGameValidator를 나누어 해결했습니다 :)

jjj0611 commented on src/main/java/racingcar/validator/MoveCountValidator.java:
```diff
@@ -0,0 +1,25 @@
+	public static void validateMoveCount(String moveCount) {
+		checkInteger(moveCount);
+		checkPositiveInteger(moveCount);
+	}
```
여기서 순서가 변경되면 어떤 부분에 대한 case를 놓칠 수 있을까요?
이런 휴먼 에러는 최소화하는 편이 좋습니다 :)

jjj0611 commented on src/main/java/racingcar/validator/MoveCountValidator.java:
```diff
@@ -0,0 +1,25 @@
+		} catch (NumberFormatException e) {
+			throw new IllegalArgumentException(ERROR_NOT_INTEGER);
+		}
+	}
```
여기서도 변환을 하고 변환이 잘 되는 것을 보고 뒤에서 변환을 하고 그 뒤에 또 시도 횟수를 만들 때 변환을 하네요. 3번이나 변환을 할 필요가 있을까요?

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+		for (String carName : carNames) {
+			this.cars.add(new Car(carName));
+		}
+	}
```
자동차의 이름이 중복되면 안된다는 로직은 어디에 존재해야 할까요? new Cars(new String[]{"j", "j"});와 같이 생성하면 문제 없이 Cars가 생성되겠네요 :)

summerlunaa replied:
그렇게는 생각하지 못했네요ㅠㅠ 위의 코멘트까지 고려해서 검증 로직을 Cars, Car로 이동시켰습니다!!

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+			.forEach(Car::moveForward);
+	}
+
+	private boolean isMoveCondition(Car car) {
```
인자로 들어온 Car는 어디에서 사용될까요?

summerlunaa replied:
앗 실수입니다 T-T 수정했습니다.

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+	private int makeRandom() {
+		Random random = new Random();
+		return random.nextInt(RANDOM_NUMBER_BOUND);
+	}
```
이 부분은 어떻게 테스트할 수 있을까요??

summerlunaa replied:
우선 private 메서드라 테스트가 힘들었는데 이 메서드는 Utils에 클래스 메서드로 들어가도 될 것 같아서 메서드 위치를 옮겼습니다. 그리고 테스트를 여러번 돌려서 0~10 사이의 숫자가 나오는지 확인해서 테스트할 수 있을 것 같습니다! 테스트 코드 추가했습니다!

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+	public List<String> getWinners() {
+		int maxPosition = getMaxPosition();
+		return this.cars.stream()
+			.filter(car -> car.getPosition() == maxPosition)
```
isSmaePosition이라는 메서드를 생성하셨는데 활용해보면 어떨까요??

jjj0611 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+		return this.cars.stream()
+			.map(Car::toString)
+			.collect(Collectors.toList());
+	}
```
method명만 보면 cars.getAllPosition()이어서 List<Position>이나 List<Integer>를 반환할 것 같은데, toString을 호출하고 있네요.

메서드 명만 보면 조금 헷갈릴 수 있을 것 같은데, 어떻게 생각하시나요?

summerlunaa replied:
저도 이 메서드 이름 지을때 고민이 많았습니다ㅠㅠ getAllPositionToString으로 한번 바꿔봤는데 어떤가요? 아직 제 마음에 들지는 않는데 더 좋은 이름이 생각나질 않습니다.

jjj0611 commented on src/test/java/calculator/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,80 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+class StringAddCalculatorTest {
```
처음으로 테스트 코드를 작성하셨다고 했는데 케이스를 잘 나누어서 작성해주셨네요 👍

jjj0611 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringAddCalculator {
```
static method라는 존재를 아시는 것 같은데, 이 부분에서는 사용하진 않았네요.
static 키워드를 학습해보시고 차이에 대해서 정리해보시면 좋을 것 같아요 :)

summerlunaa replied:
StringAddCalculator는 Util의 성격이 강해서 클래스 메서드를 사용하는 것이 맞는 것 같네요! 마침 오늘 수업시간에 관련 내용을 배워서 좋았습니다ㅎㅎ

jjj0611 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,53 @@
+		String[] splitString = split(givenString);
+		int[] intValues = stringToInt(splitString);
+		Arrays.stream(intValues).forEach(this::checkNegativeInteger);
+		return Arrays.stream(intValues).sum();
```
Stream을 2번 열어서 합계를 하고 있는데, Stream을 사용하신 김에 한 번만 열어서 합계를 해보면 어떨까요?

summerlunaa replied:
제가 아직 스트림이 익숙하지 않아서ㅠㅠ forEach, sum 둘 다 최종연산인데 어떻게 한 줄로 합칠 수 있을까요? 찾아본다고 찾아봤는데 방법을 찾지 못해서.. 우선은 수정하지 못했습니다 T-T

jjj0611 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,53 @@
+	}
+
+	private boolean checkNullOrEmpty(String givenString) {
+		return givenString == null || givenString.isEmpty();
```
null을 먼저 체크하여 givenString.isEmpty()에 NPE가 나지 않도록 한 것이 매우 좋네요 👍

jjj0611 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,53 @@
+		return givenString.split("[,:]");
+	}
+
+	private int[] stringToInt(String[] values) {
```
배열을 사용하신 이유가 있을까요? 
List와는 어떤 차이가 있나요?
다른 코드에서는 List를 사용하신 것 같은데 List로 변경해보면 어떨까요?

summerlunaa replied:
코멘트를 보고 고민해보니 제가 구현하면서 array와 List를 큰 고민 없이 집히는대로(?) 쓴 것 같아서 반성하게 되었습니다... 열심히 두 자료구조의 차이를 찾아보니 결론적으로 접근할 일이 많으면 array, 삽입, 삭제할 일이 많으면 list가 효율적이라고 하더라구요. 그런데 프리코스 때 받았던 피드백에서는 배열 대신 java collection을 사용하라고 하셔서, 특별한 경우가 아니면 list를 사용을 지향하는 것이 좋을까요? 궁금합니다!

jjj0611 replied:
엇 접근할 일이 많을 때 배열을 사용하라는건 어떤 의도였을까요??
한번 ArrayList의 구현체를 들어가보시겠어요?!

jjj0611 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,34 @@
+package racingcar.domain;
+
+public class Car {
```
car를 domain 패키지에 넣어주셨네요. 한 번 고민해봅시다.
new Car(""); new Car(null)과 같이 생성하면 어떻게 될까요?
어떻게 되는 것이 맞을까요?

summerlunaa replied:
음 제가 구현한 코드로는 원하지 않는 결과가 발생하겠네요. 자동차 이름에 대한 검증을 생성자 안에서 진행해서 "", null이 들어오는 경우에 exception이 발생하도록 했습니다. 이렇게 해결하면 될까요!?

jjj0611 replied:
넵 파랑!

검증 로직 같은 것들이 어디서 이루어져야 하는가는 쉽지 않은 문제에요.
예를 들어 아이디는 5글자 이상이어야 한다라고 했을 때, 5글자 이상이어야하는건
아이디를 입력하는 측에서 막아야할까요? 아이디가 만들어지는 지점에서 막아야 할까요?
만약 전자라면 다른 경로를 통해서 아이디가 생성될 때는 5글자 미만이어도 생성이 되는 일이 벌어질 거에요!
그래서 이게 정말로 입력과 관련된 로직이라면 입력을 처리하면서 검증해야하는 것이 맞겠지만,
좀 더 핵심로직(흔히 도메인이라고 이야기 하는)이라면 안 쪽에서 검증되는 것이 맞을 것 같아요 :)

jjj0611 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,34 @@
+		return position;
+	}
+
+	@Override
+	public String toString() {
+		return this.name + STRING_JOIN_DELIMITER + MOVE_EXPRESSION.repeat(Math.max(0, this.position));
+	}
```
자동차의 상태를 출력하기 위한 메서드로 toString을 사용하는 것이 그 목적과 부합할까요?
https://docs.oracle.com/javase/10/docs/api/java/lang/Object.html#toString() 오라클의 문서에 toString에 대한 정의를 참고해보면 좋을 것 같아요.

만약 추후에 사용자에게 자동차의 위치를 보여주는 view가 console이 아닌 다른 view를 채택하게 될 땐 어떻게 될까요?

summerlunaa replied:
흠.. 정의를 읽어봤는데 toString 메서드는 사람이 읽기 쉽도록 간결하지만 객체의 정보를 잘 나타낼 수 있는 문자열을 반환해야 하네요! 자동차의 이름과 위치가 나타나니 객체의 정보를 나타내고 있는 것 같기는 한데,,, 다른 view를 채택하게 됐을 때 toString을 수정하거나 출력하는 함수에서 변형해서 쓰는 것은 비효율적인가요?

jjj0611 replied:
엇 만약 다른 view를 채택하는 것이 아니라 두 view를 동시에 사용 가능하도록 하게 되면 어떻게 될까요??
toString이 어떤 view에 맞춰져야할까요??

summerlunaa replied:
이렇게 설명해주시니까 확 와닿습니다ㅠㅠ 2단계에서 수정하도록 할게요! 감사합니다ㅎㅎ

jjj0611 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,53 @@
+	private int intValueOf(String value) {
+		try {
+			return Integer.parseInt(value);
+		} catch (NumberFormatException e) {
```
구체적인 Exception을 잡아주는 것은 매우 좋은 습관입니다! 👍

jjj0611 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,53 @@
+		try {
+			return Integer.parseInt(value);
+		} catch (NumberFormatException e) {
+			throw new RuntimeException("숫자값을 입력해주세요");
```
RuntimeException으로 변경하여 던지는 이유가 있을까요?
RuntimeException과 Exception은 어떤 차이가 있을까요?
또 java에서 정의하는 기본 Exception들에는 어떤 것이 있을까요?

summerlunaa replied:
요구사항에 `RuntimeException 예외를 throw한다.`고 적혀있어서 별 생각 없이 RuntimeException으로 변경하여 던졌는데 제이님 덕분에 RuntimeException과 checkedException을 다시 한 번 공부하게 되었습니다! NumberFormatException이 RuntimeException에 포함되니까 굳이 더 큰 범위의 RuntimeException으로 변경하여 던질 필요가 없어 보여서 수정했습니다!!

jjj0611 replied:
넵! 다른 예외로 변화하여 던지는 것 자체가 문제가 되진 않지만,
NumberFormatException으로 잡은 예외를 RuntimeException으로 변환할 필요는 없어 보여요 :)

Java에 기본적으로 있는 예외들에 대해 찾아서 학습한 후에 좀 더 명확한 예외로 변환하는 것도 좋을 것 같아요 :)

jjj0611 commented on src/test/java/junitstudy/StringTest.java:
```diff
@@ -0,0 +1,40 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+public class StringTest {
```
StringTest와 SetTest와 같은 테스트를 학습테스트라고 합니다.
테스트를 처음 작성하신 것인데 위와 같은 테스트를 작성했다니 대단하네요👍
학습테스트를 작성하심으로 제가 얻은 효용은 앞으로 제가 볼 코드에서 어떤 기능들이 사용되었을지에 대한 감을 얻을 수 있었어요.

이처럼 테스트는 코드를 보는 다른 개발자에게 좋은 가이드라인이 되어 줍니다!

jjj0611 reviewed (changes requested):
안녕하세요 파랑! 리뷰어 제이입니다.

먼저 제 코드리뷰는 절대값이 아니고, 참고할만한 코멘트 정도로 이해해주시면 될 것 같아요.
따라서 이해가 되지 않는 부분도 있을거고 동의하지 않는 부분도 있을 수 있는데,
같이 이야기 하며 좋은 방향을 찾아가보면 좋을 것 같아요 :)

---

질문 주신 것에 대한 이야기를 먼저 하자면,

1. 랜덤값에 따라서 테스트를 짜는데 어려움을 느낀 부분

 - 혹시 어떤 부분들을 시도해보셨을까요? 파랑이 시도한 것을 바탕으로 이야기하면 더 좋을 것 같아요 :)

2. 테스트만을 위한 메서드를 작성하는 것

 - 어떤 메서드가 그런 일을 했을까요? 테스트를 하기 위해서 왜 그 메서드가 필요했는지도 알려주시면 더 좋은 논의를 할 수 있을 것 같아요 :)

3. 패키지 안 클래스가 하나여도 패키지를 만드는 것이 좋을까?

 - 어떤 패키지명과 어떤 클래스였나요? 왜 해당 패키지를 만드셨을까요?

4. SplitAndTrim에 대한 논의

 - 메서드를 작게 나누라고 해는데 해당 메서드는 두 가지 일을 하고 있네요. 메서드를 더 작게 나누면 어떨까요?

위 내용들에 대한 답변을 기반으로 이야기 하면 더 깊은 논의를 할 수 있지 않을까요? :)

---

테스트가 처음이라고 하셨는데 학습테스트부터 아주 잘 작성을 해주셨네요.
parameterized test와 setup 메서드 등 테스트 작성을 용이하게 해주는 도구들도 아주 잘 이용하신 점이 매우 인상적입니다 😄

기능 요구사항도 매우 예쁘게 잘 정리해주셨네요 markdown 문법과 친숙하신 것 같네요!

commit 로그도 아주 잘 쪼개서 나누어주셨지만, 조금 아쉬운 로그들도 보였습니다.
패키지 구조 변경이라는 커밋이 있었는데, 조금 더 구체적으로 남겼으면 어땠을까요?
물론 커밋의 변경사항만으로 의도를 파악할 수 있는 것이 베스트이지만
해당 커밋에는 패키지 구조 변경 뿐 아니라 new Cars()를 하는 부분에서의 로직도 조금 변경된 것 같아요.
메시지를 조금 더 상세하게, 메시지만 읽고도 쭉 흐름을 파악할 수 있도록 로그를 남기면 좋을 것 같아요! :)

메서드 라인이나 else와 같은 문법을 사용하지 않는 부분들은 대체적으로 잘 지켜주셨네요 👍

코멘트 드린 부분 확인해보시고 혹시 이해가 가지 않는 부분들이 있다면 말씀해주세요 :)
첫 미션 고생하셨어요!

summerlunaa committed:
style: 불필요한 bracket 제거

summerlunaa committed:
refactor: StringTest split 메서드 쪼갬

summerlunaa committed:
refactor: 자동차 이름을 split 해주는 코드의 중복을 없앰

summerlunaa committed:
refactor: NameValidator를 CarNameValidator와 RacingGameValidator로 분리

summerlunaa committed:
refactor: 검증 로직을 InputView에서 Car, Cars로 이동

summerlunaa committed:
refactor: strToInt 메서드를 Utils로 이동

summerlunaa committed:
refactor: 중복되는 정수 변환 로직 제거

summerlunaa committed:
fix: 불필요한 인자 제거

summerlunaa committed:
refactor: isSamePosition 메서드 사용

summerlunaa committed:
refactor: StringAddCalculator의 메서드 static으로 변경

summerlunaa committed:
refactor: array를 list로 변경

summerlunaa committed:
refactor: 던지는 에러 종류를 더 구체적으로 변경

summerlunaa committed:
refactor: 메서드 순서 변경

summerlunaa committed:
refactor: domain 패키지로 테스트 코드 이동

summerlunaa committed:
refactor: makeRandom 메서드를 Utils 클래스로 이동

summerlunaa committed:
test: Utils 클래스에 대한 Test 코드 작성

summerlunaa committed:
chore: 주석 코드 삭제

summerlunaa committed:
refactor: 메서드 이름 변경

summerlunaa committed:
chore: 사용하지 않는 import 삭제

summerlunaa commented:
리뷰 반영이 늦었습니다ㅠ_ㅠ 각각의 코멘트에 대해서 대부분 댓글을 달긴 했는데, 크게 변경된 부분은 아래와 같습니다!

1. Car 객체의 이름 검증, moveCount에 대한 검증 로직을 각 객체의 생성자로 이동했습니다!
    
    이렇게 해서 문자열을 나누는 작업과, 정수로 변환하는 작업을 한 번만 하도록 줄일 수 있었습니다.
    
    그리고 Car(” “) 이런식으로 객체를 생성하는 경우에도 오류가 나지 않도록 할 수 있었습니다.
    
2. NameValidator를 CarNameValidator와 RacingGameValidator로 나누었습니다!
    
    Validator를 두 가지로 나누어 CarNameValidator는 Car 생성자에서 호출하여 하나의 이름만 검증하도록 했습니다. 그리고 RacingGameValidator에서는 레이싱 경기의 입장에서 자동차가 하나밖에 없지는 않는지, 자동차의 이름이 중복되지는 않는지를 검증하도록 했습니다.
    
3. 몇몇 메서드를 Utils 클래스로 옮기고 이에 대한 테스트 코드를 추가했습니다.
    
    이렇게 해서 private라 테스트가 힘들었던 makeRandom 메서드도 범위 확인을 통해 테스트할 수 있도록 하였습니다.
    

그리고 아래 내용들을 공부했습니다.

1. Array VS List 두 자료구조의 차이
2. toString의 목적
3. RuntimeException VS checkedException 두 예외의 차이

추가적으로 처음 드렸던 질문에 대해서 말씀드리겠습니다!

1. 랜덤값에 따라서 테스트를 짜는데 어려움을 느낀 부분
    
    이번 수정에서 makeRandom 메서드를 범위 확인을 통해 테스트할 수 있도록 구현했습니다. 그런데 뭔가 부족하지 않은가 하는 기분이 듭니다ㅠㅠ. makeRandom 메서드 자체는 테스트를 했는데 Random 값 때문에 GameController는 아예 테스트를 손도 못 댔는데 이게 맞나?하는 고민이 생기네요..! test 코드에 대한 감이 아직 없는 것 같아요 ;_; 사실 제가 뭘 모르고 있는지도 잘 모르겠어요ㅎㅎㅎ ㅜㅜ
    
2. 테스트만을 위한 메서드를 작성하는 것
    
    CarsTest의 carGenerate 테스트를 위해 getSize라는 메서드를 만들었습니다! n개의 자동차 이름을 넣었을 때 n개의 자동차가 cars 리스트에 들어가 있는지를 테스트하기 위해 만들었습니다. 그런데 getSize는 다른 곳에서는 쓰이지 않는데, 이런식으로 테스트를 위한 메서드를 만들어도 되는지 궁금합니다!
    
3. 패키지 안 클래스가 하나여도 패키지를 만드는 것이 좋을까?
    
    GameController와 Utils 클래스는 따로 패키지 않에 넣지 않았는데 각각 Controller, Utils 패키지를 만들어서 넣는 것이 좋을까요?
    
4. SplitAndTrim에 대한 논의
    
    split과 trim 메서드를 따로 나누면 각각은 원래 존재하는 메서드라서 고민입니다. 그럼 splitAndTrim 메서드를 삭제하고 그냥 split 메서드 한 번, trim 메서드 한 번 각각 적용시키는 것이 나을까요?
    
    ```java
    // 현재
    public static String[] getCarNames(Scanner scanner) {
    		System.out.println(REQUEST_CAR_NAMES);
    		String inputNames = scanner.nextLine();
    		return Utils.splitAndTrim(inputNames);
    }
    
    // split, trim 따로??
    public static String[] getCarNames(Scanner scanner) {
    		System.out.println(REQUEST_CAR_NAMES);
    		String inputNames = scanner.nextLine();
    		String[] splitInputNames = inputNames.split(",");
    		Arrays.setAll(splitNames, i -> splitNames[i].trim());
    		return splitInputNames;
    }
    ```
    

마지막으로 코멘트에서도 남겼던 질문들을 다시 쓰자면

1. 두 자료구조의 차이를 찾아보니 결론적으로 접근할 일이 많으면 array, 삽입, 삭제할 일이 많으면 list가 효율적이라고 하더라구요. 그런데 프리코스 때 받았던 피드백에서는 배열 대신 java collection을 사용하라고 하셔서, 특별한 경우가 아니면 list를 사용을 지향하는 것이 좋을까요?
2. toString에 대한 정의를 읽어봤는데 toString 메서드는 사람이 읽기 쉽도록 간결하지만 객체의 정보를 잘 나타낼 수 있는 문자열을 반환해야 하네요! 자동차의 이름과 위치가 나타나니 객체의 정보를 나타내고 있는 것 같기는 한데,,, 다른 view를 채택하게 됐을 때 toString을 수정하거나 출력하는 함수에서 변형해서 쓰는 것은 비효율적인가요?

이정도인 것 같습니다. 꼼꼼히 리뷰해주셔서 감사합니다! 앞으로 더 열심히 리팩토링 해보겠습니다. 파이팅!!😊

jjj0611 reviewed (approved):
안녕하세요 파랑! 리뷰어 제이입니다.

말씀드린 것들에 대해서 너무 꼼꼼히 잘 학습하셔서 너무 좋아요! :)
또 어떤 부분들이 변경되었는지 남겨주신 것도 너무 좋네요!

리뷰를 반영하면서 생기셨던 질문들에 대해서는 제가 다시 코멘트로 달아 두었으니 한 번 확인 부탁드려요!

첫 요청 때 남겨주신 질문들에 대해서 제가 다시 질문 한 것에 답변 드린 것을 기반으로 코멘트를 달자면,

1. 뭘 모르는지 모르는 것을 아는 것이 시작입니다! Random 값 때문에 테스트하기 어려웠다면 그 값을 랜덤하지 않는 상태로 테스트할 수 있지 않을까요?? 테스트라는 것은 통제된 환경이 매우 중요하고, 그럼 그 값을 랜덤하지 않게 만들면 되지 않을까요?!
2. 저는 테스트만을 위한 메서드를 만드는 것을 선호하지는 않습니다! 만약 테스트만을 위해서 무언가가 만들어져야 한다면 설계를 다시 한 번 고민해보는 것도 좋은 방법이에요!
3. 가독성을 높이면 저는 좋다고 생각합니다. 이건 사람마다 취향이 좀 갈릴 것 같은데요. 저는 패키지도 결국 코드를 잘 관리하기 위한 방법 중 하나라고 생각합니다. 가독성을 높일 수 있고, 시스템에 대한 이해를 도울 수 있다면 좋다고 생각해요 :)
4. 이것도 가독의 차이라고 생각하긴 해요! 그런데 조금 더 좋은 이름을 지을 순 있을 것 같아요! 그리고 그 메서드를 사용하는 사용자는 내부 구현 로직에 대한 것은 조금은 몰라도 되게 숨길 수 있지 않을까요?

2단계 미션 진행을 위해 리뷰 반영해주신 것에 대해서 다 코멘트를 달지는 못했어요.
파랑이 질문 주신 내용들과 코멘트 남겨준 것에 대해서 답장만 남긴 상태인데,
반영된 코드는 2단계 제출하셨을 때 한 번에 확인하도록 할게요!

1단계 수고하셨고, 2단계 미션 진행해주세요 :)
